system/jlib/jsecrets.cpp

```diff
@@ -368,17 +369,61 @@ static bool isEmptyTimeval(const timeval &tv)
 
 //---------------------------------------------------------------------------------------------------------------------
 
+//The secret key has the form category/name[@vaultId][#version]
+
+static std::string buildSecretKey(const char * category, const char * name, const char * optVaultId, const char * optVersion)
+{
+    std::string key;
+    key.append(category).append("/").append(name);
+    if (optVaultId)
+        key.append("@").append(optVaultId);
+    if (optVersion)
+        key.append("#").append(optVersion);
+    return key;
+}
+
+static void expandSecretKey(std::string & category, std::string & name, std::string & optVaultId, std::string & optVersion, const char * key)
+{
+    const char * slash = strchr(key, '/');
+    assertex(slash);
+    const char * at = strchr(slash, '@');
```

Comment:
Will go wrong if anyone put a # into a vault id. Perhaps that is checked elsewhere

Reply:
Agree.  The vault id is just the name of the vault config entry in values.yaml or environment.xml.  I don't think there is currently any checking for allowed characters.  We have control over what is allowed and probably should limit the characters used.  Could add checking to the helm chart and the CVault constructor.

Reply:
Added HPCC-30124 to check them.